cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java

```diff
@@ -192,8 +208,38 @@ public LoggingContext getLoggingContext() {
     return loggingContext;
   }
 
+  //---- following are methods to manage transaction lifecycle for the datasets. This needs to
+  //---- be refactored after [TEPHRA-99] and [CDAP-3893] are resolved.
+
+  private void initializeTransactionAwares() {
+    // initially the tx-awares of the dataset cache are only the static datasets
+    for (TransactionAware txAware : getDatasetCache().getTransactionAwares()) {
+      txAwares.add(txAware);
+      txAware.startTx(transaction);
+    }
+  }
+
+  /**
+   * This delegates the instantiation of the dataset to the super class, but in addition, if
+   * the dataset is a transaction-aware, it starts the transaction and remembers the dataset.
+   */
+  @Override
+  public <T extends Dataset> T getDataset(String name, Map<String, String> arguments)
+    throws DatasetInstantiationException {
+    T dataset = super.getDataset(name, arguments);
+    if (dataset instanceof TransactionAware) {
+      TransactionAware txAware = (TransactionAware) dataset;
+      txAwares.add(txAware);
+      txAware.startTx(transaction);
```

Comment:
Do you need to check if the `txAware` is already in the `txAwares` set (by the result of the `add()` method) before calling `startTx`? Otherwise `startTx` might get called multiple times on the same instance?